### PR TITLE
Store GC stats in RRDs, render on client-side with C3

### PR DESCRIPTION
**Pull request by djs55**
Depends on: [ocaml/opam-repository#4313]
Depends on: [ocaml/opam-repository#4314]

src/stats.ml renders a page which references javascript built
from stats/main.ml via js_of_ocaml.

The stats/main.ml fetches updates in RRD export format from
the URI /rrd_updates, in the same way as [xapi-project/xen-api].

A background thread in src/stats.ml feeds Gc.stats() data into
RRDs via [xapi-project/xcp-rrd].

The stats/main.ml runs in the browser and draws/animates graphs
using OCaml bindings to the C3 library, built on top of D3.

Signed-off-by: David Scott <dave.scott@citrix.com>

rrd

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -22,7 +22,7 @@ MIRAGE ?= mirage
 MODE   ?= unix
 FS     ?= fat
 NET    ?= direct
-IPADDR ?= static
+IPADDR ?= dhcp
```
Comment (djs55) [inline Makefile]: oops, I should remove this

File: src/stats.ml [hunk 2]
```diff
@@ -1,53 +1,119 @@
...
+    <a href="$str:uri$">$str:t.Rrd_timescales.name$</a>
+    >>
+  ) timescales in
+  <:html<
+    <head>
+      <meta charset="utf-8" />
+      <link href="https://cdnjs.cloudflare.com/ajax/libs/c3/0.4.10/c3.css" rel="stylesheet" type="text/css"/>
+      <script src="https://cdnjs.cloudflare.com/ajax/libs/d3/3.5.5/d3.min.js" charset="utf-8"></script>
+      <script src="https://cdnjs.cloudflare.com/ajax/libs/c3/0.4.10/c3.min.js"></script>
+      <script src="/js/stats/main.js"> </script>
```
Comment (djs55) [inline src/stats.ml]: Note the external references here. I could add C3 and D3 to this repo if you'd prefer.

File: stats/main.ml [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+  req##onreadystatechange <- Js.wrap_callback
+    (fun _ ->
+       (match req##readyState with
+                   | XmlHttpRequest.DONE ->
+                           Lwt.wakeup w (Js.to_string req##responseText)
+                   | _ -> ()));
+
+	req##send (Js.some (Js.string ""));
+  Lwt.on_cancel res (fun () -> req##abort ()) ;
+  res
```
Comment (Drup) [inline stats/main.ml]: `XmlHttpRequest.get` ?

File: stats/main.ml [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+
+
+  loop (-window + 1)
+
+let _ =
+  Dom_html.window##onload <- Dom_html.handler
+    (fun _ ->
+      Lwt.async watch_rrds;
+      Js._true
+    )
```
Comment (Drup) [inline stats/main.ml]:
```ocaml
Lwt_js_events.onload >>= (fun _ -> watch_rrds ())
```

File: stats/main.ml [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+    Uri.make ~scheme:"http" ~path:"/rrd_updates" ~query () in
+
+  let window = Rrd_timescales.to_span timescale in
+
+  let rec loop start =
+    do_get ~uri:(uri start)
+    >>= fun txt ->
+    let input = Xmlm.make_input (`String (0, txt)) in
+    let update = Rrd_updates.of_xml input in
+    Firebug.console##log(Js.string "got some updates");
```
Comment (Drup) [inline stats/main.ml]: You might want to use `Lwt_log_js` instead of firebug in general. it's just plain nicer.

File: stats/Makefile [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+depend::
+	$(OPAM) install $(PKGS) --verbose
+
+main.byte:
+	$(BUILD) main.byte
+
+build: main.js
+
+main.js: main.byte
+	js_of_ocaml main.byte
```
Comment (Drup) [inline stats/Makefile]: You probably want to call jsoo_minify too.

**Comment by Drup** [comment]
Nice, It's so pretty! :3

**Comment by samoht** [comment]
That looks pretty! Let me know when this is ready to go live! (I've restarted the failing jobs on Travis)

**Comment by djs55** [comment]
```
File "dispatch_tls.ml", line 61, characters 23-34:
Error: The module DS is a functor, not a structure
Command exited with code 2.
```

**Comment by djs55** [comment]
@samoht I think it's ready!

**Comment by djs55** [comment]
Ok, I've added a second chart which shows HTTP requests/sec and errors/sec. Travis took a bit of prodding but is happy again.

**Comment by samoht** [comment]
Sweet, merging now!

**Comment by djs55** [comment]
Thanks!

**Comment by samoht** [comment]
Seems that there is a problem somewhere: 

```
GET https://mirage.io/js/stats/main.js 404 (Not Found)
```

**Comment by djs55** [comment]
hm, that ought to have been packed inside the filesystem -- maybe this isn't enough?

https://github.com/mirage/mirage-www/blob/master/stats/Makefile#L22

**Comment by djs55** [comment]
Oh, ordering problem:
```
mirage.io   Generating: /home/travis/build/mirage/mirage-www/src/static2.ml
mirage.io   => ocaml-crunch -o static2.ml /home/travis/build/mirage/mirage-www/src/../files
mirage.io   Generating: /home/travis/build/mirage/mirage-www/src/static1.ml
mirage.io   => ocaml-crunch -o static1.ml /home/travis/build/mirage/mirage-www/src/../tmpl
+make build
cd stats && make all
make[1]: Entering directory `/home/travis/build/mirage/mirage-www/stats'
ocamlbuild -use-ocamlfind -pkgs lwt.syntax,js_of_ocaml,js_of_ocaml.syntax,js_of_ocaml.log,c3,uri,re.str,rrd,bigarray -tags "syntax(camlp4o),annot,bin_annot,strict_sequence,principal,thread" -tag-line "<static*.*>: -syntax(camlp4o)" -cflag -g -lflags -g,-linkpkg main.byte
Finished, 3 targets (0 cached) in 00:00:00.
js_of_ocaml +weak.js main.byte
jsoo_minify main.js
mkdir -p ../files/js/stats
cp main.min.js ../files/js/stats/main.js
```
-- I think I need to build the js before the `mirage configure` or perhaps run `mirage configure` a second time?

**Comment by Drup** [comment]
It doesn't work in https in firefox (mixed content ...)

**Comment by djs55** [comment]
Yes indeed -- candidate fix in #349... waiting for travis to sanity-check
the build.

On Sun, Jun 28, 2015 at 11:06 PM, Gabriel Radanne <notifications@github.com>
wrote:

> It doesn't work in https in firefox (mixed content ...)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mirage/mirage-www/pull/342#issuecomment-116354430>.
>



-- 
Dave Scott

**Pull request merged**